iotdb/src/main/java/org/apache/iotdb/db/writelog/transfer/CodecInstances.java

```diff
@@ -0,0 +1,402 @@
+    ThreadLocal<ByteBuffer> localBuffer = new ThreadLocal<>();
+
+    @Override
+    public byte[] encode(InsertPlan plan) {
+      int type = SystemLogOperator.INSERT;
+      if (localBuffer.get() == null) {
+        localBuffer.set(ByteBuffer.allocate(config.getMaxLogEntrySize()));
+      }
+      ByteBuffer buffer = localBuffer.get();
+      buffer.clear();
+      buffer.put((byte) type);
+      buffer.put((byte) plan.getInsertType());
+      buffer.putLong(plan.getTime());
+
+      ByteBufferUtils.putString(buffer, plan.getDeviceId());
+
+      List<String> measurementList = plan.getMeasurements();
+      buffer.putInt(measurementList.size());
+      for (String m : measurementList) {
+        ByteBufferUtils.putString(buffer, m);
+      }
+
+      List<String> valueList = plan.getValues();
+      buffer.putInt(valueList.size());
+      for (String m : valueList) {
+        ByteBufferUtils.putString(buffer, m);
+      }
+
+      return Arrays.copyOfRange(buffer.array(), 0, buffer.position());
+    }
+
+    @Override
+    public InsertPlan decode(byte[] bytes) throws IOException {
+      ByteBuffer buffer = ByteBuffer.wrap(bytes);
+
+      buffer.get(); // read and skip an int representing "type"
+      int insertType = buffer.get();
+      long time = buffer.getLong();
+
+      String device = ByteBufferUtils.readString(buffer);
+
+      int mmListLength = buffer.getInt();
+      List<String> measurementsList = new ArrayList<>(mmListLength);
+      for (int i = 0; i < mmListLength; i++) {
+        measurementsList.add(ByteBufferUtils.readString(buffer));
+      }
+
+      int valueListLength = buffer.getInt();
+      List<String> valuesList = new ArrayList<>(valueListLength);
+      for (int i = 0; i < valueListLength; i++) {
+        valuesList.add(ByteBufferUtils.readString(buffer));
+      }
+
+      InsertPlan ans = new InsertPlan(device, time, measurementsList, valuesList);
+      ans.setInsertType(insertType);
+      return ans;
+    }
+  };
+
+  static final Codec<MetadataPlan> metadataPlanCodec = new Codec<MetadataPlan>() {
+    ThreadLocal<ByteBuffer> localBuffer = new ThreadLocal<>();
+
+    @Override
+    public byte[] encode(MetadataPlan plan) {
+      int type = SystemLogOperator.METADATA;
+      if (localBuffer.get() == null) {
+        localBuffer.set(ByteBuffer.allocate(config.getMaxLogEntrySize()));
+      }
+      ByteBuffer buffer = localBuffer.get();
+      buffer.clear();
+      buffer.put((byte) type);
+      buffer.put((byte) plan.getNamespaceType().serialize());
+      buffer.put((byte) plan.getDataType().serialize());
+      buffer.put((byte) plan.getCompressor().serialize());
+      buffer.put((byte) plan.getEncoding().serialize());
+
+      String path = plan.getPath().toString();
+      ByteBufferUtils.putString(buffer, path);
+
+      List<Path> deletePathList = plan.getDeletePathList();
+      if (deletePathList == null) {
+        buffer.putInt(-1);
```

Comment:
In this case, I like use 0 rather than an error code (i.e., -1 in this case),
then the deserialize function also needs  be adapted.

Reply:
but in this case, a new List<> will be generated in the deserialization function, though there is no deletePaths...

Reply:
I prefer -1 to 0, because -1 doesn't mean an error in this case. -1 means null and 0 means the list is empty. So I think it's better to use -1